npm/pkg/dataplane/policies/policymanager_windows.go

```diff
@@ -120,44 +132,62 @@ func (pMgr *PolicyManager) removePolicy(policy *NPMNetworkPolicy, endpointList m
 	// but if the bug is not solved then get all existing policies and remove relevant policies from list
 	// then apply remaining policies onto the endpoint
 	var aggregateErr error
+	noOfRulesToRemove := len(rulesToRemove)
 	for epIPAddr, epID := range endpointList {
-		epObj, err := pMgr.getEndpointByID(epID)
+		err := pMgr.removePolicyByEndpointID(rulesToRemove[0].Id, epID, noOfRulesToRemove, removeOnlyGivenPolicy)
 		if err != nil {
-			// Do not return if one endpoint fails, try all endpoints.
-			// aggregate the error message and return it at the end
 			aggregateErr = fmt.Errorf("[DataPlane Windows] Skipping removing policies on %s ID Endpoint with %s err\n Previous %w", epID, err.Error(), aggregateErr)
 			continue
 		}
-		if len(epObj.Policies) == 0 {
-			klog.Infof("[DataPlanewindows] No Policies to remove on %s ID Endpoint", epID)
-			continue
-		}
 
-		epBuilder, err := splitEndpointPolicies(epObj.Policies)
-		if err != nil {
-			aggregateErr = fmt.Errorf("[DataPlane Windows] Skipping removing policies on %s ID Endpoint with %s err\n Previous %w", epID, err.Error(), aggregateErr)
-			continue
-		}
+		// Delete podendpoint from policy cache
+		delete(policy.PodEndpoints, epIPAddr)
+	}
 
-		epBuilder.compareAndRemovePolicies(rulesToRemove[0].Id, len(rulesToRemove))
-		klog.Infof("[DataPlanewindows] Epbuilder ACL policies before removing %+v", epBuilder.aclPolicies)
-		klog.Infof("[DataPlanewindows] Epbuilder Other policies before removing %+v", epBuilder.otherPolicies)
-		epPolicies, err := epBuilder.getHCNPolicyRequest()
-		if err != nil {
-			aggregateErr = fmt.Errorf("[DataPlanewindows] Skipping removing policies on %s ID Endpoint with %s err\n Previous %w", epID, err.Error(), aggregateErr)
-			continue
-		}
+	return aggregateErr
+}
 
-		err = pMgr.updatePoliciesOnEndpoint(epObj, epPolicies)
-		if err != nil {
-			aggregateErr = fmt.Errorf("[DataPlanewindows] Skipping removing policies on %s ID Endpoint with %s err\n Previous %w", epID, err.Error(), aggregateErr)
-			continue
-		}
+func (pMgr *PolicyManager) removePolicyByEndpointID(ruleID, epID string, noOfRulesToRemove int, resetAllACL shouldResetAllACLs) error {
+	epObj, err := pMgr.getEndpointByID(epID)
+	var aggregateErr error
+	if err != nil {
+		// Do not return if one endpoint fails, try all endpoints.
+		// aggregate the error message and return it at the end
+		aggregateErr = fmt.Errorf("[DataPlane Windows] Skipping removing policies on %s ID Endpoint with %s err\n Previous %w", epID, err.Error(), aggregateErr)
```

Comment:
shouldn't we just return errors immediately now?

Reply:
Agreed, correcting this.